src/Bicep.Core/Emit/TemplateWriter.cs

```diff
@@ -1018,6 +1025,75 @@ private void EmitProviders(ExpressionEmitter emitter, ImmutableArray<DeclaredPro
             });
         }
 
+        private void EmitProvidersV2(ExpressionEmitter emitter, ImmutableArray<DeclaredProviderExpression> providers)
+        {
+            emitter.EmitObjectProperty("extensions", () =>
+            {
+                foreach (var provider in providers)
+                {
+                    var settings = provider.Settings;
+
+                    emitter.EmitObjectProperty(provider.Name, () =>
+                    {
+                        emitter.EmitProperty("name", settings.ArmTemplateProviderName);
+                        emitter.EmitProperty("version", settings.ArmTemplateProviderVersion);
+
+                        EmitProviderV2Config(provider, emitter);
+                    },
+                    provider.SourceSyntax);
+                }
+            });
+        }
+
+        private void EmitProviderV2Config(DeclaredProviderExpression provider, ExpressionEmitter emitter)
+        {
+            if (provider.Config is null)
+            {
+                return;
+            }
+
+            if (provider.Config is not ObjectExpression providerConfig)
+            {
+                throw new UnreachableException($"Provider config type expected to be of type: '{nameof(ObjectExpression)}' but received: '{provider.Config.GetType()}'");
+            }
+
+            emitter.EmitObjectProperty("config", () =>
+            {
+                foreach (var providerConfigProperty in providerConfig.Properties)
+                {
+                    // Type checking should have validated that the config name is not an expression (e.g. string interpolation), if we get a null value it means something
+                    // was wrong with type checking validation.
+                    var providerConfigName = providerConfigProperty.TryGetKeyText() ?? throw new UnreachableException("Expressions are not allowed as config names.");
+                    var x = provider.Settings.ConfigurationType ?? throw new UnreachableException();
+                    var providerConfigType = GetProviderConfigType(providerConfigName, provider.Settings.ConfigurationType!);
```

Comment:
Did you mean to use the `x` variable rather than `provider.Settings.ConfigurationType!`?